src/security/KeyData.cc

```diff
@@ -78,13 +79,51 @@ Security::KeyData::loadX509CertFromFile()
     return bool(cert);
 }
 
+#if USE_OPENSSL || USE_GNUTLS
+static void
+verifyChainCert(Security::CertPointer &ca, Security::CertPointer &latestCert, Security::CertList &chain)
+{
+    auto nameStr = Security::CertSubjectName(ca);
+    Security::ErrorCode checkCode;
+#if !USE_OPENSSL || TLS_CHAIN_NO_SELFSIGNED
```

Comment:
I find the 2-condition clause I used simpler and more direct than the nested 3-clause you are proposing.

Reply:
> I find the 2-condition clause I used simpler and more direct 

Unfortunately, for the PR "2-condition clause" to actually work and make sense, it has to be interpreted as the following complex 4(?) condition clause that the compiler sees:

```C++
#if (USE_OPENSSL || USE_GNUTLS) && (!USE_OPENSSL || TLS_CHAIN_NO_SELFSIGNED)
```

In other words, yes, the PR condition looks "simpler" on the surface, but that simplicity is actually misleading. Correctly interpreting that shorter expression is difficult and gives (or should give) most uninitiated code readers pause. It took me a minute to understand what it _really_ means.

My condition has more characters, but it removes the implicit dependency on the surrounding `USE_OPENSSL || USE_GNUTLS` precondition, removes difficult-to-interpret-correctly negation of `USE_OPENSSL`, and is easy to read -- what you see is exactly what the condition means:

```C++
#if USE_GNUTLS || (USE_OPENSSL && TLS_CHAIN_NO_SELFSIGNED)
```

Please note that my conversion assumes that `USE_GNUTLS` implies `!USE_OPENSSL`. IIRC, we have agreed that this implication is always true. If my recollection is wrong, then the actual condition has a different independent form which raises questions about the correctness of the condition itself:

```C++
#if (USE_GNUTLS && !USE_OPENSSL) || (USE_GNUTLS && TLS_CHAIN_NO_SELFSIGNED) || (USE_OPENSSL && TLS_CHAIN_NO_SELFSIGNED)
```

Reply:
The thing is this TLS_CHAIN_NO_SELFSIGNED option is one of our custom ones. It was intended to permit OpenSSL builds to revert to the old behaviour of sending self-signed certs.
This self-signed check was added to fix problems with Node.js and GnuTLS clients which reject validation of chains containing the root CA.

Since this PR is trying for parity the proper form should not have included the library-specific clause at all. Thank you for persisting about this.